core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -88,134 +50,326 @@ class Queue[A] private (
    */
-  final val shutdown: UIO[Unit] = (for {
-    hook <- shutdownHook.modify {
-             case None       => (IO.unit, None)
-             case Some(hook) => (hook, None)
-           }
-    takers <- IO.effectTotal(unsafePollAll(takers))
-    _      <- IO.foreachPar(takers)(_.interrupt) *> hook
-    _      <- strategy.shutdown
-  } yield ()).uninterruptible
+  def shutdown: IO[E, Unit]
 
   /**
    * Removes the oldest value in the queue. If the queue is empty, this will
    * return a computation that resumes when an item has been added to the queue.
    */
-  final val take: UIO[A] =
-    for {
-      _ <- checkShutdownState
-
-      item <- IO.effectTotalWith { platform =>
-               val item = queue.poll(null.asInstanceOf[A])
-               if (item != null) strategy.unsafeOnQueueEmptySpace(queue, platform)
-               item
-             }
-
-      a <- if (item != null) IO.succeedLazy(item)
-          else
-            for {
-              p <- Promise.make[Nothing, A]
-              // add the promise to takers, then:
-              // - try take again in case a value was added since
-              // - wait for the promise to be completed
-              // - clean up resources in case of interruption
-              a <- (IO.effectTotalWith { platform =>
-                    takers.offer(p)
-                    unsafeCompleteTakers(platform)
-                  } *> p.await).onInterrupt(removeTaker(p))
-            } yield a
-    } yield a
+  def take: IO[E, B]
 
   /**
    * Removes all the values in the queue and returns the list of the values. If the queue
    * is empty returns empty list.
    */
-  final val takeAll: UIO[List[A]] =
-    for {
-      _ <- checkShutdownState
-
-      as <- IO.effectTotalWith { platform =>
-             val as = unsafePollAll(queue)
-             strategy.unsafeOnQueueEmptySpace(queue, platform)
-             as
-           }
-    } yield as
+  def takeAll: IO[E, List[B]]
 
   /**
    * Takes up to max number of values in the queue.
    */
-  final def takeUpTo(max: Int): UIO[List[A]] =
-    for {
-      _ <- checkShutdownState
+  def takeUpTo(max: Int): IO[E, List[B]]
+
+  def map[C](f: B => C): Queue2[E, A, C] =
+    new Queue2[E, A, C] {
+      def capacity: Int                             = self.capacity
+      def offer(a: A): IO[E, Boolean]               = self.offer(a)
+      def offerAll(as: Iterable[A]): IO[E, Boolean] = self.offerAll(as)
+      def awaitShutdown: IO[E, Unit]                = self.awaitShutdown
+      def size: IO[E, Int]                          = self.size
+      def shutdown: IO[E, Unit]                     = self.shutdown
+      def take: IO[E, C]                            = self.take.map(f)
+      def takeAll: IO[E, List[C]]                   = self.takeAll.map(_.map(f))
+      def takeUpTo(max: Int): IO[E, List[C]]        = self.takeUpTo(max).map(_.map(f))
+    }
+
+  def mapM[E2 >: E, C](f: B => IO[E2, C]): Queue2[E2, A, C] =
```

Comment:
The existence of these `M` methods, with the added type parameter `R`, makes me wonder if we should add `R` like stream so you can pull in environment as you're doing effects. Like with stream it's not absolutely required since you can thread it in yourself from a higher level. But if we can do it at no cost to type inference, makes sense to be consistent with Stream. Queue is like a push stream really.